### PR TITLE
Origin localization server

### PR DESCRIPTION
**Pull request by lexious89**
## Description
Localization algorithm for positioning and pathloss computation (rssi analytics) implementation in server.
## Solved issue(s)
#56 #7

### CODE REVIEW

**Comment by platisd** [comment]
What issue does this pull request solve and what does it do in general?

File: get_dependencies.sh [hunk 1]
```diff
@@ -13,5 +13,7 @@ mkdir build && cd build
 cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release ..
 make
 sudo make install
+
+sudo apt-get install libxml2-dev
```
Comment (platisd) [inline get_dependencies.sh:17]: Since we have this as a dependency, do we need to also pull the `libxml` source in in the `external` folder?

File: server/CMakeLists.txt [hunk 2]
```diff
@@ -1,22 +1,31 @@
-cmake_minimum_required(VERSION 3.2)
+cmake_minimum_required(VERSION 2.6)
```
Comment (platisd) [inline server/CMakeLists.txt]: Why downgrade?

File: server/CMakeLists.txt [hunk 3]
```diff
@@ -1,22 +1,31 @@
...
 
 include_directories(
     ${CMAKE_FIND_ROOT_PATH}/include
     ${CMAKE_CURRENT_SOURCE_DIR}/include
-    ${CMAKE_CURRENT_SOURCE_DIR}/include/external )
+    ${CMAKE_CURRENT_SOURCE_DIR}/include/external)
+
+link_directories(${CMAKE_CURRENT_SOURCE_DIR}/libxml2)
 
-link_directories(${CMAKE_FIND_ROOT_PATH}/lib)
```
Comment (platisd) [inline server/CMakeLists.txt]: What was `lib` doing before and now it's not necessary?

File: server/CMakeLists.txt [hunk 4]
```diff
@@ -1,22 +1,31 @@
...
+link_directories(${CMAKE_FIND_ROOT_PATH}/libxml2)
 add_executable(ins_server ${SOURCE_FILES})
-target_link_libraries(ins_server pistache.a sqlite3.a dl)
+target_link_libraries(ins_server pistache.a sqlite3.a dl m ${LIBXML2_LIBRARIES})
```
Comment (platisd) [inline server/CMakeLists.txt]: > m ${LIBXML2_LIBRARIES}

Where is the `LIBXML2_LIBRARIES` set and what does that `m` do in the front of it?

File: server/README.md [hunk 5]
```diff
@@ -20,6 +20,10 @@ The INS server provides a backend RESTful service for INS-node and front-end web
     `http://localhost:9080/set_rssi/5239/23:43:3d:3e:5e:f5/13/5a:4e:44:ff:5a:6e/40`
     `http://localhost:9080/set_rssi/5239/23:43:3d:3e:5e:f5/11.3/5a:4e:44:ff:5a:6e/44.3`
     `http://localhost:9080/set_rssi/5239/23:43:3d:3e:5e:f5/15.6/5a:4e:44:ff:5a:6e/44.4`
+    
+    http://localhost:9080/set_rssi/5239/23:43:3d:3e:5e:f5/-53/23:43:3d:3e:5e:f6/-52/23:43:3d:3e:5e:f7/-20/23:43:3d:3e:5e:f5/-53/23:43:3d:3e:5e:f6/-52/23:43:3d:3e:5e:f7/-20/23:43:3d:3e:5e:f5/-59/23:43:3d:3e:5e:f6/-51/23:43:3d:3e:5e:f7/-22
+     http://localhost:9080/set_rssi/5239/23:43:3d:3e:5e:f5/-43/23:43:3d:3e:5e:f6/-55/23:43:3d:3e:5e:f7/-18
+      http://localhost:9080/set_rssi/5239/23:43:3d:3e:5e:f5/-59/23:43:3d:3e:5e:f6/-51/23:43:3d:3e:5e:f7/-22
```
Comment (platisd) [inline server/README.md]: Why do we need these additional lines in the documentation? In any case they should be formatted appropriately.

File: server/README.md [hunk 6]
```diff
@@ -31,7 +35,7 @@ The INS server provides a backend RESTful service for INS-node and front-end web
     #### Example
   * INS-node with device id 1000 triggers device position computation after sending several RSSI readings.
 
-    `http://localhost:9080/resolve_pos/1000`
+    `http://localhost:9080/resolve_pos/5239`
```
Comment (platisd) [inline server/README.md]: What was wrong with `id 1000`? :stuck_out_tongue_closed_eyes:

File: server/WifiNodeLCFG.xml [hunk 7]
```diff
@@ -0,0 +1,505 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (platisd) [inline server/WifiNodeLCFG.xml:1]: What does this file do? Please enter some description as a comment inside it. Is this something we should have version controlled?

File: server/include/external/libxml/DOCBparser.h [hunk 8]
```diff
@@ -0,0 +1,96 @@
+/*
+ * Summary: old DocBook SGML parser
+ * Description: interface for a DocBook SGML non-verifying parser
+ * This code is DEPRECATED, and should not be used anymore.
```
Comment (platisd) [inline server/include/external/libxml/DOCBparser.h]: Are we using this code? If so, this file says we shouldn't! :laughing:

File: server/include/ins_service.hpp [hunk 9]
```diff
@@ -13,6 +13,11 @@
 #include "data_store.hpp"
 #include "localization.hpp"
 #include "types.hpp"
+extern "C"
+{
+#include <WifiNode.h>
+#include <WifiAccessPointLocalConfig.h>
+}
```
Comment (platisd) [inline server/include/ins_service.hpp:20]: It is going to be a nightmare to unit tests this. :cry:

File: server/include/localization.hpp [hunk 10]
```diff
@@ -20,22 +27,26 @@ class Localization
 {
 public:
 #ifdef ENABLE_TESTS
-    friend class LocalizationFixture;
+	friend class LocalizationFixture;
```
Comment (platisd) [inline server/include/localization.hpp:30]: Please let's be civil here and use spaces instead of tabs! :grin:

Comment (platisd) [reply server/include/localization.hpp:30]: Same goes for the rest of this file.

File: server/src/ResolveLocation.c [hunk 11]
```diff
@@ -0,0 +1,114 @@
+/*
+ * kalman.c
+ *
+ *  Created on: 23 nov 2017
+ *      Author: alexs
+ */
```
Comment (platisd) [inline server/src/ResolveLocation.c]: Remove comment header, it looks outdated anyway :smiley:

File: server/src/ResolveLocation.c [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+
+#include <math.h>
+#include <stdint.h>
+#include "../include/ResolveLocation.h"
```
Comment (platisd) [inline server/src/ResolveLocation.c]: Can we include this in a different way without specifying the relative path (i.e. include it via the CMake)?

File: server/src/ResolveLocation.c [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+	return (rssi + powerCalibrator);
+}
+
+
+float kalmanProcess(float rssiBuffer[NUMBER_SAMPLES])
+{
+	float Estimate[NUMBER_SAMPLES+1];
+	float kalmanGain = 0;
+	float initialErrorEstimate = INITIAL_ERROR_ESTIMATE;
+	//float initialMeasurement;
```
Comment (platisd) [inline server/src/ResolveLocation.c]: Remove commented out code.

File: server/src/ResolveLocation.c [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+
+
+float power2distance(float Power_d)
+{
+	return (distance_o * pow((POWER_do - Power_d)/(10 * n_factor),10));
+}
+
+
+uint32_t trilateration_process(cord_t* nodePosition, struct co_ordinate_distance_tag trilaterationParams[TRILATERAT_NUMBER_NODES])
```
Comment (platisd) [inline server/src/ResolveLocation.c]: If we are supposed to return a `uint32_t` where is the `return` statement? :smile:

**Comment by lexious89** [comment]
I have updated the description and solved issues.

File: server/src/ResolveLocation.c [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+		trilaterationParams[i].distance	                 = wifiNodeParams[i]->cordinatesDistance.distance;
+
+		for (uint32_t j = 0 ; j < CORDINATE_DIMENSION ; j++)
+		{
+			trilaterationParams[i].position.co_ord[j] = wifiNodeParams[i]->cordinatesDistance.position.co_ord[j];
+		}
+
+	}
+
+	trilateration_process(nodePosition, trilaterationParams);
```
Comment (platisd) [inline server/src/ResolveLocation.c]: Why not have the `trilateration_process` return a `cord_t *` so you can return it directly? This way you can also skip passing it as an argument all together.

File: server/src/WifiAccessPointLocalConfig.c [hunk 16]
```diff
@@ -0,0 +1,244 @@
...
+void print_element_names(xmlNode * a_node)
+{
+    xmlNode *cur_node = NULL;
+
+    for (cur_node = a_node; cur_node; cur_node = cur_node->next) {
+        if (cur_node->type == XML_ELEMENT_NODE) {
+            printf("node type: Element, name: %s\n", cur_node->name);
+        }
+
+        print_element_names(cur_node->children);
```
Comment (platisd) [inline server/src/WifiAccessPointLocalConfig.c]: Is this a recursion?! Cool, but isn't it dangerous to do such in C?

Also, why do we need this `print_element_names` function if no one is using it?

**Comment by lexious89** [comment]
I also made some 'cosmetic changes' in the server/readme.md and some other files. These changes are trivial and only helped with debugging the localization alg. You may ignore these changes. They really don't add any value to the sw except for debugging purposes.

File: server/src/data_store.cpp [hunk 17]
```diff
@@ -268,9 +270,9 @@ std::vector<AccessPoint> DataStore::GetDistinctAccessPoints(const std::string& d
     return access_points;
 }
 
-std::vector<int32_t> DataStore::GetRSSISeriesData(const std::string& device_id, AccessPoint access_point)
+std::vector<int32_t> DataStore::GetRSSISeriesFromDatabase(const std::string& device_id, AccessPoint access_point)
 {
-    console_->debug(" + DataStore::GetRSSISeriesData");
+    //console_->debug(" + DataStore::GetRSSISeriesDatabase");
```
Comment (platisd) [inline server/src/data_store.cpp]: Why comment this out?

Comment (platisd) [reply server/src/data_store.cpp]: Same with the one below in this file.

File: server/src/localization.cpp [hunk 18]
```diff
@@ -3,17 +3,94 @@
...
 
 #include "localization.hpp"
+#include "data_store.hpp"
+
+extern insNode_t * insNoderoot;
 
 namespace ins_service
 {
 
+wifiParams_t * Localization::FillNodeDataPoints(wifiParams_t *  wifiNodeBlock, AccessPointRssiListPair mac_rssi_)
```
Comment (platisd) [inline server/src/localization.cpp]: I don't feel comfortable at all playing around with raw pointers in C++. I don't think they should be used at all.

File: server/src/localization.cpp [hunk 19]
```diff
@@ -3,17 +3,94 @@
...
+{
+	initKalmanParams(wifiNodeBlock);
+
+	wifiNodeBlock->macAddress = (char *) mac_rssi_.first.mac_addr.c_str();
```
Comment (platisd) [inline server/src/localization.cpp]: C-style casts should be avoided in C++. :cry:

File: server/src/localization.cpp [hunk 20]
```diff
@@ -3,17 +3,94 @@
...
+
+	if ( (insNode = findWifiNode(insNoderoot, device_id)) == NULL)
+	{
+		insNode =  createInsNodeListDevice(device_id); // check to make sure nodeblock exits!!
+	}
+
+	std::string msg = "Number of distinct mac addresses received from device: " + std::string(device_id) + " is: " + std::to_string(mac_rssi_list.size()) + " with: " +  std::to_string(mac_rssi_list[0].second.size()) + " data points.";
+	console_->debug(msg);
+
+	for (int i = 0; i < mac_rssi_list.size(); ++i)
```
Comment (platisd) [inline server/src/localization.cpp]: Iterate through the vector instead such as:
`for (auto mac_rssi : mac_rssi_list)`

File: server/test/suite_localization.cpp [hunk 21]
```diff
@@ -17,10 +19,113 @@ class LocalizationFixture : public ::testing::Test
...
 	Localization localization_;
+    std::shared_ptr<DataStore> data_store_ = std::make_shared<DataStore>();
 };
 
-TEST_F(LocalizationFixture, Init)
+/*
+ * TEST: Localize
+ * EXPECT: <x,y,z> of dummy node.
+*/
+TEST_F(LocalizationFixture, RoundTripLocalizationTest)
```
Comment (platisd) [inline server/test/suite_localization.cpp]: Can you specify what you are testing and how? You can use a naming convention such as `SoftwareUnderTest_WhenSomethingHappens_WillPerformAction`.

**Review by platisd — CHANGES_REQUESTED** [review]
Great work overall! Looks like something that'd work. And kudos for adding some tests at least! :grinning: 

A **LOT** of code, which makes it very challenging to review. I would like to raise the following concerns aside from my other review comments:

* **Documentation**
  * At least on the function level, I would like to see what each function does, otherwise it is extremely hard to follow.
* **Formatting**
  * The formatting throughout the code and compared to the rest of the project is not consistent. Please don't use tabs at all and if possible clang-format your code.
* **Use of C language**
  * I would like to avoid it but if needed be I can accept it.
* **Use of raw pointers in C++ code**
  * This is a red line for me. They shouldn't be used when writing C++. :smiling_imp: 

Overall looks like it works but I cannot follow 100% and I doubt I would be able even if there was adequate documentation. Can we arrange a code walk-through?

File: get_dependencies.sh (hunk 1 above)
Comment (lexious89) [reply get_dependencies.sh:17]: No, we would not need to do that. Just the header files which we have in external/. The .so package will be pulled with the above command.

File: server/CMakeLists.txt (hunk 2 above)
Comment (lexious89) [reply server/CMakeLists.txt]: Actually not a necessary downgrade. A Forgotten debug trail...

File: server/CMakeLists.txt (hunk 3 above)
Comment (lexious89) [reply server/CMakeLists.txt]: It was not necessary before. And will actually not be necessary now. That directory did not exist.

File: server/CMakeLists.txt (hunk 4 above)
Comment (lexious89) [reply server/CMakeLists.txt]: it is set from the fetch package command. The m is a math library.

File: server/README.md (hunk 5 above)
Comment (lexious89) [reply server/README.md]: Like I mentioned in the comments before, these are trivial lines I used for debugging the code. Its not a necessary change but helped when I wanted to send three rssi per command for the localization to work.

File: server/README.md (hunk 6 above)
Comment (lexious89) [reply server/README.md]: same as before. Just keeping the commands easy for copy and paste during debugging.

File: server/WifiNodeLCFG.xml (hunk 7 above)
Comment (lexious89) [reply server/WifiNodeLCFG.xml:1]: This is the localconfig file which contains the wifi positions and transmit power levels. It will be used when tuning the system. During the trilateration, the server comes here and fetches all necessary paramaters to compute distance and node position. We could leave it as it is, and not version control it. It any changes to it will be private anyway.

File: server/include/external/libxml/DOCBparser.h (hunk 8 above)
Comment (lexious89) [reply server/include/external/libxml/DOCBparser.h]: Yeh.. i am not sure about this. There are several header files that are not used at the moment but are part of the libxm2 library package. My fear is that removing such files may contain necessary variables that will be needed by the library in the future.
There could be another way with the CMakeLists file. I can look into that.

File: server/include/ins_service.hpp (hunk 9 above)
Comment (lexious89) [reply server/include/ins_service.hpp:20]: :smile:  why?

File: server/include/localization.hpp (hunk 10 above)
Comment (lexious89) [reply server/include/localization.hpp:30]: ...tabs are man's best friend. :grinning:

File: server/src/ResolveLocation.c (hunk 11 above)
Comment (lexious89) [reply server/src/ResolveLocation.c]: the whole file is actually. I will remove this.

File: server/src/ResolveLocation.c (hunk 12 above)
Comment (lexious89) [reply server/src/ResolveLocation.c]: same goes for this.

File: server/src/ResolveLocation.c (hunk 13 above)
Comment (lexious89) [reply server/src/ResolveLocation.c]: :grin:

File: server/src/ResolveLocation.c (hunk 14 above)
Comment (lexious89) [reply server/src/ResolveLocation.c]: :grin:

File: server/src/WifiAccessPointLocalConfig.c (hunk 16 above)
Comment (lexious89) [reply server/src/WifiAccessPointLocalConfig.c]: Yeh, it is not used. Iterating through all the leaves of a tree can be a daunting task. I only included it for the future. But as for recursion being dangerous, it could be because of infinite recursion possibilities. But in this case, there is a for loop which terminates when a cur_node is null. That makes it amicable.

File: server/src/data_store.cpp (hunk 17 above)
Comment (lexious89) [reply server/src/data_store.cpp]: No need. Debug trails.

File: server/src/localization.cpp (hunk 18 above)
Comment (lexious89) [reply server/src/localization.cpp]: Why?

File: server/src/localization.cpp (hunk 19 above)
Comment (lexious89) [reply server/src/localization.cpp]: Actually a cast is not needed here. Will take this out.

File: server/src/localization.cpp (hunk 20 above)
Comment (lexious89) [reply server/src/localization.cpp]: Yeh, I thought so too. But when things were not working, I went low level. :smile:

File: server/test/suite_localization.cpp (hunk 21 above)
Comment (lexious89) [reply server/test/suite_localization.cpp]: yes, sure.

**Comment by lexious89** [comment]
I understand that is a lot of code. I doubt if I would remember everything myself in a few months. A walk though would be good!

File: server/CMakeLists.txt (hunk 4 above)
Comment (lexious89) [reply server/CMakeLists.txt]: The ${LIBXML2_INCLUDE_DIR} actually worked for the main build and the test build. So I have removed the libxml2 header path from external/.
I think we can make it work for the spdlog library too with something like ${SPDLOG_INCLUDE_DIR} and ${SPDLOG_LIBRARIES}. @samup4web what do you think about that?

**Comment by platisd** [comment]
Is this ready to be reviewed again?

**Comment by lexious89** [comment]
Yes, it is.

File: server/README.md (hunk 5 above)
Comment (platisd) [reply server/README.md]: I really think this should be removed! :laughing:

File: server/README.md (hunk 6 above)
Comment (platisd) [reply server/README.md]: If you must keep it as `5239` then please update the documentation two lines above!

**Review by platisd — CHANGES_REQUESTED** [review]
OK, let's say like this, if you can run clang-format over the code changes and address the documentation issues then I can agree that this should go in.

Overall I don't feel that I have done a good code review and that we are missing a lot of documentation here, but for the sake of moving the project forward I will turn a blind eye! :grin:
What do you think?

**Review by platisd — DISMISSED** [review]
(ﾉ◕ヮ◕)ﾉ*:･ﾟ✧ ✧ﾟ･: *ヽ(◕ヮ◕ヽ)

**Review by platisd — APPROVED** [review]
(ﾉ◕ヮ◕)ﾉ*:･ﾟ✧ ✧ﾟ･: *ヽ(◕ヮ◕ヽ)

**Comment by platisd** [comment]
If more commits are to be added let me know when the request will be ready for review. Looks good now!

**Comment by lexious89** [comment]
Okej. This should be the final commit for this branch. It is ready for review. ;)
